src/us/kbase/auth2/lib/storage/mongo/MongoStorage.java

```diff
@@ -748,33 +748,49 @@ private Document getUserDoc(
 			final boolean local)
 			throws AuthStorageException, NoSuchUserException {
 		requireNonNull(userName, "userName");
-		Document user = getUserDocWithoutPassword(collection, userName);
+		return getUserDoc(
+				collection, new Document(Fields.USER_NAME, userName.getName()), local, userName);
+		
+	}
+	
+	private Document getUserDoc(
+			final String collection,
+			final Document query,
+			final boolean local,
+			final UserName nameForException)
+			throws AuthStorageException, NoSuchUserException {
+		Document user = getUserDocWithoutPassword(collection, query);
 		if (user == null) {
-			throw new NoSuchUserException(userName.getName());
+			if (nameForException != null) {
+				throw new NoSuchUserException(nameForException.getName());
+			} else {
+				return null;
+			}
 		}
+		final String userName = user.getString(Fields.USER_NAME);
 		if (user.getString(Fields.USER_ANONYMOUS_ID) == null) {
 			user = updateUserAnonID(collection, userName);
 		}
 		if (local && !user.getBoolean(Fields.USER_LOCAL)) {
-			throw new NoSuchLocalUserException(userName.getName());
+			throw new NoSuchLocalUserException(userName);
 		}
 		return user;
 	}
 
-	private Document getUserDocWithoutPassword(final String collection, final UserName userName)
+	private Document getUserDocWithoutPassword(final String collection, final Document query)
 			throws AuthStorageException {
 		return findOne(
 				collection,
-				new Document(Fields.USER_NAME, userName.getName()),
+				query,
 				new Document(Fields.USER_PWD_HSH, 0).append(Fields.USER_SALT, 0));
 	}
 
-	private Document updateUserAnonID(final String collection, final UserName userName)
+	private Document updateUserAnonID(final String collection, final String userName)
```

Comment:
Why should `userName` be a String here instead of a UserName? I think you would ensure it's really a UserName before the change.

Reply:
This is an internal method only so I'm a lot less strict about classes, etc, as I assume it's getting called by other methods in the same module and the programmers know what they're doing. In this case, the method is only called by the `getUserDoc` method, which just fetched the user document from mongo so we can assume the user name is ok. Wrapping it in a `UserName` class seems a bit pointless since it's unwrapped immediately in this method.